### ISSUE TITLE
changes to git workflow

### ISSUE THREAD

**Issue by robkooper** [PecanProject/pecan#1346]
I would like to propose a small change to the way we do branching and merging things. I have been doing this on other projects and it helps with releases.

We will have 2 branches in PEcAn that will always be there:
- master : this is the branch that holds the latest released version of PEcAn. You can pull this to your local server and have the latest version of PEcAn running. This branch will only be updated when we have a release.
- develop : this branch contains the latest and greatest code. All pull requests should go to this branch. Any code changes you make should start from this branch, and be merged using a pull request back into this branch.

This means that the workflow for git changes just a little:
```
git checkout develop
git pull
git checkout -b <issuenumber>-description
# do lots of cool work
git commit
git push origin <issuenumber>-description
# either more work and commits/push
# finally pull request
```

The only change from the original workflow is that we now start from develop, instead of master.

When it is time for a release we will first create a branch from the develop branch called `release/vnumber`. We will update at this point the version number in the release branch. At this point any fixes that need to be made to get the release working correctly are made to this release branch. Any other development work that is not part of this release will happen as before on the develop branch. Advantage is that we do not have to worry about any new code sneaking into a release code base that is not yet tested, and people can continue with developing cool things in the develop branch that can potentially break a release.

Once we are finished with the release branch and think it is ready, we merge it both into the master and develop branch (to make sure any fixes are in develop). 

Once this is done we have officially released a new version of PEcAn. Any site can now pull the latest version from the master branch, compile and have a new version of PEcAn up and running.

I would like to get some feedback from all developers to see if this makes sense and if this seems like a good idea.

**Comment by tonygardella** [comment]
Awesome! I was researching best methods for this and [this](http://nvie.com/posts/a-successful-git-branching-model/) was one way of doing things I liked and it's essentially the same. I've put it in today's discussion topics for the telecon.

**Comment by ashiklom** [comment]
I don't think this is a good idea. I think this is a great idea. I might even argue for three branches:

* `master` -- the latest stable release
* `nightly` -- where we pull in new features once they have been tested; this is the version that should run on pecan2 and similar servers.
* `develop` -- where we pull in code that has only been minimally tested (Travis, etc.). This is the version that people should keep up-to-date on their personal testing environments (e.g. VM, test-pecan).

Basically, we would submit pull requests to `develop`, which, after several independent integration tests by multiple users are merged into `nightly` (say, once a week or so), and then, on release day, are merged into `master`. This may be too complicated, but I think it's a good way of getting pull requests pulled quickly without a fear of breaking things.

**Comment by mdietze** [comment]
I'm good with the develop / master / release system proposed by @robkooper and nicely explained in the article @tonygardella linked to.  I think the idea of release branches really solves the problem I was worried about with having to get a STABLE debugged version of PEcAn ready for the flux course while at the same time having an influx of new summer developers (plus existing developers having more time to code), which would tend to lead to instability.

For collaborative feature branches, I'm mostly in support, but I'm a bit nervous that this will create PR that are larger, which I really struggle with. So I'd support this part too if we try to keep to our current pattern of many small PR rather than single large PR, but just switched to devel instead of master. 

Less convinced by @ashiklom proposal for a third 'nightly' (but really weekly) branch. It seems to accentuate one of the things that I'm already a bit hesitant about, which is the ability to 'cherry pick' which new features in the devel branch we want to go into the next version. Basically, it's not possible (or at least easy) to cherry pick, so you're taking all or nothing, so you're mixing the new parts that are stabilizing (fixing issues) with the parts that are breaking things. I worry that trying to make Alexey's idea work would mean that we'd only be able to do new PR on Monday and then the only other PR into devel the rest of the week would be debugging, since you wouldn't want to do a new PR on Friday right before devel gets pulled into nightly/weekly. These issues still happen with the release branch approach, but they happen on a ~10 week cycle rather than every single week.  Let's try @robkooper version and then see if we see an advantage of adding another layer.

**Comment by dlebauer** [comment]
Would make sense to set develop as the default branch when we implement this? https://help.github.com/articles/setting-the-default-branch/

**Comment by robkooper** [comment]
After yesterdays discussion, I have created a 'develop' branch, and made it the default target like @dlebauer said. Please if you have outstanding pull requests change the target branch to be develop.

**Comment by robkooper** [comment]
@tonygardella can you update the documentation for the git workflow?

**Comment by tonygardella** [comment]
@robkooper On it!

**Comment by infotroph** [comment]
The documentation now has a section covering the current workflow, but also still has the older material alongside it -- those should probably be deleted to avoid confusion.

**Comment by robkooper** [comment]
Added ot season of docs as one of the tasks/ideas.

**Comment by Aariq** [comment]
Looks like the current documentation is consistent with these suggestions https://pecanproject.github.io/pecan-documentation/develop/pecan-git.html
